web/api/v1/langserver.go

```diff
@@ -0,0 +1,134 @@
+// Copyright 2020 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v1
+
+import (
+	"context"
+	"net/http"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"github.com/prometheus-community/promql-langserver/prometheus"
+	"github.com/prometheus-community/promql-langserver/rest"
+	"github.com/prometheus/client_golang/api/prometheus/v1"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/pkg/labels"
+	"github.com/prometheus/prometheus/storage"
+)
+
+func NewLangServer(externalURL string, q storage.Queryable,
+	tr func(context.Context) TargetRetriever, logger log.Logger) (http.Handler, error) {
+	// create the custom prometheus client dedicated to the langserver
+	langserverClient := newPrometheusLangServerClient(externalURL, newWebkit(q, tr))
+	// TODO use the method CreateInstHandler once it is taking the interface prometheus.Registerer in parameter
+	// instead of the struct Registry
+	return rest.CreateHandler(context.Background(), langserverClient, logger)
+}
+
+type prometheusLangServerClient struct {
+	prometheus.Client
+	externalURL string
+	webkit      *webkit
+}
+
+func newPrometheusLangServerClient(externalURL string, webkit *webkit) prometheus.Client {
+	return &prometheusLangServerClient{
+		externalURL: externalURL,
+		webkit:      webkit,
+	}
+}
+
+// Metadata returns the first occurrence of metadata about metrics currently scraped by the metric name.
+func (c *prometheusLangServerClient) Metadata(ctx context.Context, metric string) (v1.Metadata, error) {
+	metadata := c.webkit.metadata(ctx, metric, 1)
+	if len(metadata) == 0 {
+		return v1.Metadata{}, nil
+	}
+	return v1.Metadata{
+		Type: v1.MetricType(metadata[metric][0].Type),
+		Help: metadata[metric][0].Help,
+		Unit: metadata[metric][0].Unit,
+	}, nil
+}
+
+// AllMetadata returns metadata about metrics currently scraped for all existing metrics.
+func (c *prometheusLangServerClient) AllMetadata(ctx context.Context) (map[string][]v1.Metadata, error) {
+	metadata := c.webkit.metadata(ctx, "", -1)
+	result := make(map[string][]v1.Metadata)
+	for metric, infos := range metadata {
```

Comment:
Can we avoid the copying here?

Reply:
I don't think that's possible. The webkit is returning the same map but with the private struct `metadata` which is different from the one coming from the public package `v1` from the go client.

The only way to avoid it so far is that the prometheus API is using the public struct `v1.Metadata` instead of the private one.

Reply:
which could make sense since that would mean the client lib and the API would share the same data model. But this something I cannot decide by my own.